nginx/lua/webdav_content.lua

```diff
@@ -0,0 +1,204 @@
+    local reader = res.body_reader
+    local buffer_size = 16*1024*1024
+    ngx.say("Beginning TPC")
+    local bytes_moved = 0
+    local last_marker_time = uv.now()
+    local adler_state = {a=1, b=0}
+    repeat
+        local buffer, err = reader(buffer_size)
+        if err then
+            ngx.log(ngx.ERR, err)
+            break
+        end
+        if buffer then
+            local current_time = uv.now()
+            if (current_time - last_marker_time > 1000) then
+              last_marker_time = current_time
+              ngx.say("TODO - add performance markers (needed for gfal) " .. current_time)
+            end
+            -- TODO: build checksum
+            -- can use LuaJIT FFI to call C function
+            -- https://stackoverflow.com/questions/53805913/how-to-define-c-functions-with-luajit
+            -- e.g. a C function for https://en.wikipedia.org/wiki/Adler-32
+            -- libz has this function
+            -- TODO: coroutine https://www.lua.org/manual/5.1/manual.html#2.11
+            adler_state = adler32_increment(adler_state, buffer) 
+            file:write(buffer)
+        end
+    until not buffer
+    ngx.say("TPC complete")
+    file:close()
+    local adler_value = adler32_finalize(adler_state)
+    ngx.say("Adler32 is " .. adler_value)
+    setxattr(destination_localpath, "user.nginx-webdav.adler32", tostring(adler_value))
+
+    -- this allows the connection to be reused by other requests
+    ok, err = httpc:set_keepalive()
+    if not ok then
+        -- TODO: is this a fatal error?
+        ngx.status = ngx.HTTP_INTERNAL_SERVER_ERROR
+        ngx.say("failed to set keepalive: ", err)
+        return ngx.exit(ngx.OK)
+    end
+end
+
+local function getxattr(path, key, value)
+  -- FIXME this doesn't work at all since we need to receive the string back
+  -- from a pointer to a void*
+  -- https://luajit.org/ext_ffi_tutorial.html
+  return ffi.C.getxattr(path, key, value, string.len(value), 0);
+end
+
+if ngx.var.request_method == "COPY" then
+    -- The COPY method is supported by ngx_http_dav_module but only for files on the same server.
+    -- We intercept the method here to support third-party push copy.
+    -- TODO: is this the best spot in the request lifecycle to do this?
+    -- https://openresty-reference.readthedocs.io/en/latest/Directives/
+    if not ngx.var.http_source then
+        ngx.status = ngx.HTTP_BAD_REQUEST
+        ngx.say("no source provided")
+        return ngx.exit(ngx.OK)
+    end
+
+    if ngx.var.http_destination then
+        ngx.status = ngx.HTTP_NOT_IMPLEMENTED
+        ngx.say("third-party push copy not implemented")
+        return ngx.exit(ngx.OK)
+    end
+
+    -- TODO: better way to find the local file location?
+    third_party_pull(ngx.var.http_source, "/var/www" .. ngx.var.request_uri)
+
+    -- At this point, the connection will either be safely back in the pool, or closed.
+    return ngx.exit(ngx.HTTP_OK)
+elseif ngx.var.request_method == "GET" then
```

Comment:
I would hope nginx handles GET already quite well, no?

Reply:
Yeah, the problem is that if you have a content handler, there's no way to say "only use for PUT and COPY", I did see a way after I posted this where you make a "fake" location in the nginx config that doesn't use the content handler, and then the GET branch here does an internal redirect thing to the fake location. I'll re-jigger that in a future version